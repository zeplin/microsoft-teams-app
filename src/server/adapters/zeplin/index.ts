export * from "./zeplin";
export * from "./types";
