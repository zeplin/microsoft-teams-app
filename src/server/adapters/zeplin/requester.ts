import axios, { AxiosInstance, AxiosRequestConfig } from "axios";
import { ZeplinError } from "./ZeplinError";
import { INTERNAL_SERVER_ERROR } from "http-status-codes";

export class Requester {
    private instance: AxiosInstance;
    constructor(config: AxiosRequestConfig) {
        this.instance = axios.create(config);
    }

    async delete(url: string, config?: AxiosRequestConfig): Promise<void> {
        try {
            await this.instance.delete(url, config);
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async get<T = object>(url: string, config?: AxiosRequestConfig): Promise<T> {
        try {
            const { data } = await this.instance.get<T>(url, config);
            return data;
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async patch(url: string, data: object, config?: AxiosRequestConfig): Promise<void> {
        try {
            await this.instance.patch(url, data, config);
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async post(url: string, data: object, config?: AxiosRequestConfig): Promise<string> {
        try {
            const { data: { id } } = await this.instance.post<{ id: string }>(url, data, config);
            return id;
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }
}
