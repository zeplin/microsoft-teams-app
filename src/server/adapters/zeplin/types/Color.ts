export interface Color {
    id: string;
    created: number;
    name: string;
    r: number;
    g: number;
    b: number;
    a: number;
}
