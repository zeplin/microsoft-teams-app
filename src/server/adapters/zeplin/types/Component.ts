import { Thumbnails } from "./Thumbnails";

export interface Component {
    id: string;
    name: string;
    created: number;
    updated?: number;
    description?: string;
    image?: {
        width: number;
        height: number;
        original_url?: string;
        thumbnails?: Thumbnails;
    };
    section?: {
        id: string;
        group?: {
            id: string;
        };
    };
}
