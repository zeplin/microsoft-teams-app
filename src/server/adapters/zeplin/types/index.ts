export * from "./Organization";
export * from "./Project";
export * from "./Styleguide";
export * from "./RemPreferences";
export * from "./WebhookEvents";
export * from "./WebhookResourceType";
export * from "./ZeplinApplication";
export * from "./ProjectWebhook";
export * from "./StyleguideWebhook";
export * from "./User";
export * from "./WebhookStatus";
export * from "./WebhookUrlHealth";
