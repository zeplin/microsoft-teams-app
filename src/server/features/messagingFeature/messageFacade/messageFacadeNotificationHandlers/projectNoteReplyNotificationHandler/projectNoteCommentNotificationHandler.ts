import { NotificationHandler } from "../NotificationHandler";
import { AdaptiveCard, commonTeamsCard } from "../teamsCardTemplates";
import {
    WebhookEvent,
    EventType,
    EventPayload,
    ProjectContext,
    ScreenContext,
    NoteContext
} from "../../../messagingTypes";
import { ScreenNoteCommentResource } from "../resources/screenNoteCommentResource";
import { ZEPLIN_MAC_APP_URL_SCHEME, ZEPLIN_WEB_APP_BASE_URL } from "../../../../../config";

type ProjectNoteCommentEventDescriptor = {
    type: EventType.PROJECT_NOTE_COMMENT;
    action: "created";
}

class ProjectNoteCommentNotificationHandler extends NotificationHandler {
    delay: 0;

    private getWebappURL(event: WebhookEvent<ProjectNoteCommentEventPayload>): string {
        const {
            payload: {
                context: {
                    screen: {
                        id: screenId
                    },
                    project: {
                        id: projectId
                    },
                    note: {
                        id: noteId
                    }
                },
                resource: {
                    id: commentId
                }
            }
        } = event;
        const webappURL = new URL(ZEPLIN_WEB_APP_BASE_URL);
        webappURL.pathname = `project/${projectId}/screen/${screenId}`;
        webappURL.searchParams.set("did", noteId);
        webappURL.searchParams.set("cmid", commentId);
        return webappURL.toString();
    }

    getGroupingKey(event: WebhookEvent): string {
        return event.deliveryId;
    }

    getTeamsMessage(events: WebhookEvent<ProjectNoteCommentEventPayload>[]): AdaptiveCard {
        if (events.length > 1) {
            // TODO: Log that there was somehow grouping done for this notification
        }

        const [event] = events;
        const {
            payload: {
                context: {
                    project: {
                        id: projectId,
                        name: projectName
                    },
                    screen: {
                        id: screenId,
                        name: screenName
                    },
                    note: {
                        id: noteId,
                        order
                    }
                },
                resource: {
                    id: commentId,
                    data: {
                        content: commentContent
                    }
                },
                actor: {
                    user: {
                        username
                    }
                }
            }
        } = event;
        const cardText = `**${username}** replied to note _#${order}_ on _${screenName}_ screen in _${projectName}_. üèÉ‚Äç‚ôÇ`;
        return commonTeamsCard({
            text: cardText,
            sectionText: commentContent,
            links: [{
                title: "Open in App",
                url: `${ZEPLIN_MAC_APP_URL_SCHEME}dot?pid=${projectId}&sid=${screenId}&did=${noteId}&cmids=${commentId}`
            }, {
                title: "Open in Web",
                url: this.getWebappURL(event)
            }]
        });
    }
}

export type ProjectNoteCommentEventPayload = EventPayload<
    ProjectNoteCommentEventDescriptor,
    ProjectContext & ScreenContext & NoteContext,
    ScreenNoteCommentResource
>;
export const projectNoteCommentNotificationHandler = new ProjectNoteCommentNotificationHandler();