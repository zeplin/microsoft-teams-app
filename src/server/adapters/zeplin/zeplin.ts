import { ProjectWebhooks } from "./ProjectWebhooks";
import { Requester } from "./requester";
import { StyleguideWebhooks } from "./StyleguideWebhooks";

interface ZeplinOptions {
    url: string;
    webhookSecret: string;
}

class Zeplin {
    public projectWebhooks!: ProjectWebhooks;
    public styleguideWebhooks!: StyleguideWebhooks;

    init({ url, webhookSecret }: ZeplinOptions): void {
        const requester = new Requester({ baseURL: `${url}/v1/` });
        this.projectWebhooks = new ProjectWebhooks(requester, webhookSecret);
        this.styleguideWebhooks = new StyleguideWebhooks(requester, webhookSecret);
    }
}

export const zeplin = new Zeplin();
