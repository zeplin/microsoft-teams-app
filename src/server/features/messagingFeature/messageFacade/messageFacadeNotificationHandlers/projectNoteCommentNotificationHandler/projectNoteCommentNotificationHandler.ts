import { NotificationHandler } from "../NotificationHandler";
import { MessageCard, commonTeamsCard } from "../teamsCardTemplates";
import { ProjectNoteCommentEvent, WebhookEvent } from "../../../../../adapters/zeplin/types";
import { ZEPLIN_MAC_APP_URL_SCHEME, ZEPLIN_WEB_APP_BASE_URL } from "../../../../../config";
import { getMacAppRedirectURL } from "../getMacAppRedirectURL";

class ProjectNoteCommentNotificationHandler extends NotificationHandler {
    delay: 0;
    private getText(event: ProjectNoteCommentEvent): string {
        const {
            payload: {
                context: {
                    project: {
                        name: projectName
                    },
                    screen: {
                        name: screenName
                    },
                    note: {
                        order
                    }
                },
                actor: {
                    user: {
                        username
                    }
                }
            }
        } = event;
        return `**${username}** replied to note _#${order}_ on _${screenName}_ screen in _${projectName}_. üèÉ‚Äç‚ôÇ`;
    }

    private getSectionText(event: ProjectNoteCommentEvent): string {
        const {
            payload: {
                resource: {
                    data: {
                        content: commentContent
                    }
                }
            }
        } = event;
        return commentContent;
    }

    private getWebappURL(event: ProjectNoteCommentEvent): string {
        const {
            payload: {
                context: {
                    screen: {
                        id: screenId
                    },
                    project: {
                        id: projectId
                    },
                    note: {
                        id: noteId
                    }
                },
                resource: {
                    id: commentId
                }
            }
        } = event;
        const webappURL = new URL(ZEPLIN_WEB_APP_BASE_URL);
        webappURL.pathname = `project/${projectId}/screen/${screenId}`;
        webappURL.searchParams.set("did", noteId);
        webappURL.searchParams.set("cmid", commentId);
        return webappURL.toString();
    }

    private getMacAppURL(event: ProjectNoteCommentEvent): string {
        const {
            payload: {
                context: {
                    project: {
                        id: projectId
                    },
                    screen: {
                        id: screenId
                    },
                    note: {
                        id: noteId
                    }
                },
                resource: {
                    id: commentId
                }
            }
        } = event;
        return getMacAppRedirectURL(`${ZEPLIN_MAC_APP_URL_SCHEME}://dot?pid=${projectId}&sid=${screenId}&did=${noteId}&cmids=${commentId}`);
    }

    shouldHandleEvent(event: WebhookEvent): boolean {
        return event.payload.action === "created";
    }

    getGroupingKey(event: WebhookEvent): string {
        return event.deliveryId;
    }

    getTeamsMessage(events: ProjectNoteCommentEvent[]): MessageCard {
        const [event] = events;
        return commonTeamsCard({
            text: this.getText(event),
            section: {
                text: this.getSectionText(event)
            },
            links: [{
                title: "Open in App",
                url: this.getMacAppURL(event)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(event)
            }]
        });
    }
}

export const projectNoteCommentNotificationHandler = new ProjectNoteCommentNotificationHandler();
