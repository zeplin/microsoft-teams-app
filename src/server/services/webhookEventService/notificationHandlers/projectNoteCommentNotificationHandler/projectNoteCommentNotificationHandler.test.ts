import { projectNoteCommentNotificationHandler } from "./projectNoteCommentNotificationHandler";
import { NoteCommentCreateEvent } from "../../../../adapters/zeplin/types";

type GetDummyEventParams = {
    projectId?: string;
    projectName?: string;
    screenId?: string;
    screenName?: string;
    username?: string;
    noteId?: string;
    noteOrder?: string;
    commentId?: string;
    commentContent?: string;
}

function getDummyEvent({
    projectId = "projectId",
    projectName = "Project MS Teams Integration",
    screenId = "screenId",
    screenName = "Manage Zeplin Connector",
    username = "sertac",
    noteId = "noteId",
    noteOrder = "1",
    commentId = "commentId",
    commentContent = "cok guzel keyfimiz var"
}: GetDummyEventParams = {}): NoteCommentCreateEvent {
    return {
        payload: {
            action: "created",
            context: {
                screen: {
                    id: screenId,
                    name: screenName
                },
                project: {
                    id: projectId,
                    name: projectName
                },
                note: {
                    id: noteId,
                    order: noteOrder
                }
            },
            actor: {
                user: {
                    username
                }
            },
            resource: {
                id: commentId,
                data: {
                    id: commentId,
                    content: commentContent
                }
            }
        }
    } as NoteCommentCreateEvent;
}

describe("projectNoteNotificationHandler", () => {
    describe("getTeamsMessage", () => {
        it("should match snapshot", () => {
            expect(projectNoteCommentNotificationHandler.getTeamsMessage([getDummyEvent()])).toMatchSnapshot();
        });
    });
});
