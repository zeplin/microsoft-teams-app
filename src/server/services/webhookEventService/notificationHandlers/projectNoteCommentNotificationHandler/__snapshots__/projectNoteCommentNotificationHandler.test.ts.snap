// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectNoteNotificationHandler getTeamsMessage should match snapshot 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**sertac** replied to note _#1_ on _Manage Zeplin Connector_ screen in _Project MS Teams Integration_. üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "cok guzel keyfimiz var",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fdot%3Fpid%3DprojectId%26sid%3DscreenId%26did%3DnoteId%26cmids%3DcommentId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/screen/screenId?did=noteId&cmid=commentId"
        }
      ]
    }
  ]
}
`;
